library/bignum_core.c

```diff
@@ -0,0 +1,296 @@
+
+/* For GCC and Clang, have builtins for byte swapping. */
+#if defined(__GNUC__) && defined(__GNUC_PREREQ)
+#if __GNUC_PREREQ(4,3)
+#define have_bswap
+#endif
+#endif
+
+#if defined(__clang__) && defined(__has_builtin)
+#if __has_builtin(__builtin_bswap32)  &&                 \
+    __has_builtin(__builtin_bswap64)
+#define have_bswap
+#endif
+#endif
+
+#if defined(have_bswap)
+    /* The compiler is hopefully able to statically evaluate this! */
+    switch( sizeof(mbedtls_mpi_uint) )
+    {
+        case 4:
+            return( __builtin_bswap32(x) );
+        case 8:
+            return( __builtin_bswap64(x) );
+    }
+#endif
+#endif /* __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__ */
+#endif /* __BYTE_ORDER__ */
+
+    /* Fall back to C-based reordering if we don't know the byte order
+     * or we couldn't use a compiler-specific builtin. */
+    return( mpi_bigendian_to_host_c( x ) );
+}
+
+void mbedtls_mpi_core_bigendian_to_host( mbedtls_mpi_uint * const X,
+                                         size_t limbs )
+{
+    mbedtls_mpi_uint *cur_limb_left;
+    mbedtls_mpi_uint *cur_limb_right;
+    if( limbs == 0 )
+        return;
+
+    /*
+     * Traverse limbs and
+     * - adapt byte-order in each limb
+     * - swap the limbs themselves.
+     * For that, simultaneously traverse the limbs from left to right
+     * and from right to left, as long as the left index is not bigger
+     * than the right index (it's not a problem if limbs is odd and the
+     * indices coincide in the last iteration).
+     */
+    for( cur_limb_left = X, cur_limb_right = X + ( limbs - 1 );
+         cur_limb_left <= cur_limb_right;
+         cur_limb_left++, cur_limb_right-- )
+    {
+        mbedtls_mpi_uint tmp;
+        /* Note that if cur_limb_left == cur_limb_right,
+         * this code effectively swaps the bytes only once. */
+        tmp             = mpi_bigendian_to_host( *cur_limb_left  );
+        *cur_limb_left  = mpi_bigendian_to_host( *cur_limb_right );
+        *cur_limb_right = tmp;
+    }
+}
+
+int mbedtls_mpi_core_read_le( mbedtls_mpi_uint *X,
+                              size_t nx,
+                              const unsigned char *buf,
+                              size_t buflen )
+{
+    size_t i;
+    size_t const limbs = CHARS_TO_LIMBS( buflen );
+
+    if( nx < limbs )
+        return( MBEDTLS_ERR_MPI_BUFFER_TOO_SMALL );
+
+    if( X != NULL )
+    {
+        memset( X, 0, nx * ciL );
+
+        for( i = 0; i < buflen; i++ )
+            X[i / ciL] |= ((mbedtls_mpi_uint) buf[i]) << ((i % ciL) << 3);
+    }
+
+    return( 0 );
+}
+
+int mbedtls_mpi_core_read_be( mbedtls_mpi_uint *X,
+                              size_t nx,
+                              const unsigned char *buf,
+                              size_t buflen )
+{
+    size_t const limbs = CHARS_TO_LIMBS( buflen );
+    size_t overhead;
+    unsigned char *Xp;
+
+    if( nx < limbs )
+        return( MBEDTLS_ERR_MPI_BUFFER_TOO_SMALL );
+
+    if( X != NULL )
```

Comment:
So rather than do it this way, how about

```
int mbedtls_mpi_core_read_be( mbedtls_mpi_uint *X,
                              size_t nx,
                              const unsigned char *buf,
                              size_t buflen )
{
    size_t const limbs = CHARS_TO_LIMBS( buflen );

    if( nx < limbs )
        return( MBEDTLS_ERR_MPI_BUFFER_TOO_SMALL );

    /* If nx is 0, buflen must also be 0 (from previous test). Nothing to do. */
    if( nx == 0 )
        return( 0 );

    memset( X, 0, nx * ciL );

    if( buflen != 0 )   /* memcpy() with (NULL, 0) is undefined behaviour */
    {
        size_t overhead = ( nx * ciL ) - buflen;
        unsigned char *Xp = (unsigned char *) X;
        memcpy( Xp + overhead, buf, buflen );
    }

    mbedtls_mpi_core_bigendian_to_host( X, nx );

    return( 0 );
}
```

which seems simpler.

The documentation should also be clear about buffers of length `0` being allowed to have a pointer of `NULL`. Ideally that would be in one place, otherwise we need doc comments such as

```
/** Import X from unsigned binary data, big endian.
 *
 * The MPI needs to have enough limbs to store the full value (including any
 * most significant zero bytes in the input).
 *
 * \param X      The address of the MPI.
 *               May only be #NULL if \nx is 0 and \p buflen is 0.
 * \param nx     The number of limbs of \p X.
 * \param buf    The input buffer to import from.
 *               May only be #NULL if \p buflen is 0.
 * \param buflen The length in bytes of \p buf.
 *
 * \return       \c 0 if successful.
 * \return       #MBEDTLS_ERR_MPI_BUFFER_TOO_SMALL if \p X isn't
 *               large enough to hold the value in \p buf.
 */
int mbedtls_mpi_core_read_be( ...
```

Reply:
(and with that, if the test still fails, it's then clearly an incorrect test, and that can be fixed!)